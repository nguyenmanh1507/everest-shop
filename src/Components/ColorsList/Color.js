// @flow

import React, { PureComponent } from 'react'

import type { ColorType } from './ColorsListType'

class Color extends PureComponent<ColorType> {
  render() {
    const { color, selectMulti, name, handleColorSelect, ...rest } = this.props

    return (
      <label className={`brick brick--${color} mr-xs`}>
        <input
          data-test="color-input"
          type={selectMulti ? 'checkbox' : 'radio'}
          name={name}
          onChange={handleColorSelect}
          value={`${name}:${color}`}
          {...rest}
        />
        <span className="brick__body" />
      </label>
    )
  }
}

export default Color
