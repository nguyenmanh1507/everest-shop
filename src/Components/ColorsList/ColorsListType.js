// @flow

type ColorSuit =
  | 'red'
  | 'pink'
  | 'purple'
  | 'deep-blue'
  | 'indigo'
  | 'blue'
  | 'light-blue'
  | 'cyan'
  | 'teal'
  | 'green'
  | 'light-green'
  | 'lime'
  | 'yellow'
  | 'amber'
  | 'orange'
  | 'deep-orange'
  | 'brow'
  | 'grey'
  | 'blue-grey'
  | 'black'

export type ColorsListType = {
  /** Function handle value on change */
  handleColorSelect: () => void,
  /** Group name for colors */
  name: string,
  /** Choose multi color or not */
  selectMulti?: boolean,
  /** Array color of choice (color name follow Material colors) */
  colors?: Array<ColorSuit>
}

export type ColorType = {
  /** Color name */
  color: string,
  /** Function handle value on change */
  handleColorSelect: () => void,
  /** Group name for colors */
  name: string,
  /** Choose multi color or not */
  selectMulti?: boolean
}
