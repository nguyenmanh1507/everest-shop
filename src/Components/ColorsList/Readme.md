ColorsList example

### Select single

```js
<ColorsList name="select-single" handleColorSelect={() => {}} />
```

### Select multi

```js
<ColorsList name="select-multi" selectMulti handleColorSelect={() => {}} />
```

### Generate colors of choice

```js
<ColorsList
  colors={['teal', 'indigo', 'green', 'deep-orange']}
  name="color-choice"
  handleColorSelect={() => {}}
/>
```
