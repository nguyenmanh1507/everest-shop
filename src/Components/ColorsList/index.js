// @flow

export { default } from './ColorsList'
