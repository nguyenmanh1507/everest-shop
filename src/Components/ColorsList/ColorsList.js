// @flow

import React, { PureComponent } from 'react'
import { uniqueId, uniq } from 'lodash'

import type { ColorsListType } from './ColorsListType'
import Color from './Color'

const colorCodeNames = [
  'red',
  'pink',
  'purple',
  'deep-blue',
  'indigo',
  'blue',
  'light-blue',
  'cyan',
  'teal',
  'green',
  'light-green',
  'lime',
  'yellow',
  'amber',
  'orange',
  'deep-orange',
  'brow',
  'grey',
  'blue-grey',
  'black'
]

class ColorsList extends PureComponent<ColorsListType> {
  static defaultProps = {
    selectMulti: false
  }

  render() {
    const { colors, selectMulti, handleColorSelect, name, ...rest } = this.props

    if (colors && colors.length > 0) {
      if (uniq(colors).length !== colors.length) {
        console.warn('Duplicate color name')
      }

      return (
        <div>
          {uniq(colors).map(color => (
            <Color
              key={uniqueId(`color-${color}-`)}
              color={color}
              selectMulti={selectMulti}
              name={name}
              handleColorSelect={handleColorSelect}
              {...rest}
            />
          ))}
        </div>
      )
    }

    return (
      <div>
        {colorCodeNames.map(color => (
          <Color
            key={uniqueId(`color-${color}-`)}
            color={color}
            selectMulti={selectMulti}
            name={name}
            handleColorSelect={handleColorSelect}
            {...rest}
          />
        ))}
      </div>
    )
  }
}

export default ColorsList
