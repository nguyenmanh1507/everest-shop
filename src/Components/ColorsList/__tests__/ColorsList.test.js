// @flow

import React from 'react'
import renderer from 'react-test-renderer'
import { mount, render, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import forEach from 'lodash/forEach'

import ColorsList from '../ColorsList'

configure({ adapter: new Adapter() })

const mockFn = jest.fn()

test('ColorsList render properly', () => {
  const component = renderer.create(
    <ColorsList name="select-single-test" handleColorSelect={() => {}} />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('ColorsList render list of checkbox', () => {
  const component = renderer.create(
    <ColorsList
      name="select-single-test"
      selectMulti
      handleColorSelect={() => {}}
    />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('ColorsList render list of choice colors', () => {
  const colors = ['teal', 'indigo', 'green', 'deep-orange']

  const component = renderer.create(
    <ColorsList
      name="select-single-test"
      colors={colors}
      handleColorSelect={() => {}}
    />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('<ColorList>', () => {
  const colorInputName = 'color'
  const wrapper = render(renderComp(colorInputName))

  test('All input should has same name attribute', () => {
    forEach(wrapper.find('[data-test="color-input"]'), node => {
      expect(node.attribs.name).toBe(colorInputName)
    })
  })

  test('Should render input radio by default', () => {
    forEach(wrapper.find('[data-test="color-input"]'), node => {
      expect(node.attribs.type).toBe('radio')
    })
  })

  test('Should render input radio by default', () => {
    const w = render(renderComp(colorInputName, true))
    forEach(w.find('[data-test="color-input"]'), node => {
      expect(node.attribs.type).toBe('checkbox')
    })
  })

  test('Should has number of input equal colors props length', () => {
    const colors = ['teal', 'indigo', 'green', 'deep-orange']
    const colorsLength = colors.length
    const w = render(renderComp(colorInputName, undefined, colors))

    expect(w.find('[data-test="color-input"]')).toHaveLength(colorsLength)
  })

  test('Should remove duplicate name in colors props', () => {
    const colors = ['teal', 'indigo', 'green', 'deep-orange']
    const colorsLength = colors.length
    const colorsDuplicate = [
      'teal',
      'teal',
      'indigo',
      'green',
      'green',
      'deep-orange'
    ]
    const w = render(renderComp(colorInputName, undefined, colorsDuplicate))

    expect(w.find('[data-test="color-input"]')).toHaveLength(colorsLength)
  })

  test('Should run handle function once when color select', () => {
    const w = mount(renderComp())
    w
      .find('[data-test="color-input"]')
      .first()
      .simulate('change')
    expect(mockFn.mock.calls.length).toBe(1)
  })
})

function renderComp(
  name = 'select-color-test',
  selectMulti = false,
  colors = [],
  cb = mockFn
) {
  return (
    <ColorsList
      name={name}
      selectMulti={selectMulti}
      colors={colors}
      handleColorSelect={cb}
    />
  )
}
