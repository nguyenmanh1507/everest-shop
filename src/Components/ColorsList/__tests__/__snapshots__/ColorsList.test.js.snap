// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorsList render list of checkbox 1`] = `
<div>
  <label
    className="brick brick--red mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:red"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--pink mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:pink"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--purple mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:purple"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--deep-blue mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:deep-blue"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--indigo mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:indigo"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--blue mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:blue"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--light-blue mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:light-blue"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--cyan mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:cyan"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--teal mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:teal"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--green mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:green"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--light-green mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:light-green"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--lime mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:lime"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--yellow mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:yellow"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--amber mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:amber"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--orange mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:orange"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--deep-orange mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:deep-orange"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--brow mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:brow"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--grey mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:grey"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--blue-grey mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:blue-grey"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--black mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:black"
    />
    <span
      className="brick__body"
    />
  </label>
</div>
`;

exports[`ColorsList render list of choice colors 1`] = `
<div>
  <label
    className="brick brick--teal mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:teal"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--indigo mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:indigo"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--green mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:green"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--deep-orange mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:deep-orange"
    />
    <span
      className="brick__body"
    />
  </label>
</div>
`;

exports[`ColorsList render properly 1`] = `
<div>
  <label
    className="brick brick--red mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:red"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--pink mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:pink"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--purple mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:purple"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--deep-blue mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:deep-blue"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--indigo mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:indigo"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--blue mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:blue"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--light-blue mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:light-blue"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--cyan mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:cyan"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--teal mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:teal"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--green mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:green"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--light-green mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:light-green"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--lime mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:lime"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--yellow mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:yellow"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--amber mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:amber"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--orange mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:orange"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--deep-orange mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:deep-orange"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--brow mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:brow"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--grey mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:grey"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--blue-grey mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:blue-grey"
    />
    <span
      className="brick__body"
    />
  </label>
  <label
    className="brick brick--black mr-xs"
  >
    <input
      data-test="color-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:black"
    />
    <span
      className="brick__body"
    />
  </label>
</div>
`;
